File: 2020/Python/01-ReportRepair/ReportRepair.py
#!/usr/bin/env python3

# Report Repair
# https://adventofcode.com/2020/day/1

def read_file(name):
    file = open(name)
    return list(file.readlines())

def find_2020(expenses):
    for i in range(len(expenses)-1):
        for j in range(i+1, len(expenses)):
            if expenses[i] + expenses[j] == 2020:
                return (expenses[i], expenses[j])

def find_three_2020(expenses):
    for i in range(len(expenses)-2):
        for j in range(i+1, len(expenses)-1):
            for k in range(j+1, len(expenses)):
                if expenses[i] + expenses[j] + expenses[k] == 2020:
                    return (expenses[i], expenses[j], expenses[k])


expenses = list(map(int, read_file('01-input.txt')))
sum_values = find_2020(expenses)
product = sum_values[0] * sum_values[1]
print("The product of the two expenses is {0}".format(product))

sum_three_values = find_three_2020(expenses)
product_three = sum_three_values[0] * sum_three_values[1] * sum_three_values[2]
print("The prdouct of the three expenses is {0}".format(product_three))

File: 2020/Swift/01-ReportRepair/ReportRepair.swift
#!/usr/bin/swift

// Report Repair
// https://adventofcode.com/2020/day/1

import Foundation

func readFile(named name: String) -> [String] {
    let currentDirectoryURL = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
    let fileURL = URL(fileURLWithPath: name + ".txt", relativeTo: currentDirectoryURL)
    guard let content = try? String(contentsOf: fileURL, encoding: String.Encoding.utf8) else {
        print("Unable to read input file \(name)")
        print("Current directory: \(currentDirectoryURL)")
        return []
    }
    return content.components(separatedBy: .newlines)
}

func find2020(expenses: [Int]) -> (Int, Int) {
    for i in 0 ..< expenses.count - 1 {
        for j in i+1 ..< expenses.count {
            if expenses[i] + expenses[j] == 2020 {
                return (expenses[i], expenses[j])
            }
        }
    }
    print("No expenses sum to 2020")
    return (0, 0)
}

func findThree2020(expenses: [Int]) -> (Int, Int, Int) {
    for i in 0 ..< expenses.count - 2 {
        for j in i+1 ..< expenses.count - 1 {
            for k in j+1 ..< expenses.count {
                if expenses[i] + expenses[j] + expenses[k] == 2020 {
                    return (expenses[i], expenses[j], expenses[k])
                }
            }
        }
    }
    print("No expenses sum to 2020")
    return (0, 0, 0)
}

let expenses = readFile(named: "01-input").compactMap { Int($0) }
let (expense1, expense2) = find2020(expenses: expenses)
let product = expense1 * expense2
print("The product of two expenses is \(product)")

let (expense3, expense4, expense5) = findThree2020(expenses: expenses)
let productThree = expense3 * expense4 * expense5
print("The product of three expenses is \(productThree)")
